src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.cs

```diff
@@ -320,8 +320,7 @@ protected BoundStatement GenerateParameterStorage(LocalSymbol stateMachineVariab
             }
 
             var builtBody = bodyBuilder.ToImmutableAndFree();
-            ImmutableArray<BoundStatement> newBody = LocalRewriter.TryConstructNullCheckedStatementList(method.Parameters, builtBody, F);
-            return newBody.IsDefault ? F.Block(builtBody) : F.Block(ImmutableArray.Create(stateMachineVariable), newBody);
```

Comment:
The ImmutableArray here specifically was the cause of the crash. newBody already contained a block with the stateMachineVariable in it, so adding it again here made us crash in code generation.